gdb: support aarch64-darwin as a hostPlatform

Feyorsh opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Thanks to the wonderful cross compilation support in nixpkgs, GDB (along with a full cross compile toolchain) can be spun up in a jiffy:
```
$ nix-shell -I nixpkgs=. -E "{pkgs ? (import <nixpkgs> { crossSystem = ''x86_64-linux''; }).__splicedPackages }: with pkgs; mkShell { packages = [ gdb ]; }"

[nix-shell:~/Personal/nixpkgs]$ x86_64-unknown-linux-gnu-gdb
GNU gdb (GDB) 14.2
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=aarch64-apple-darwin --target=x86_64-unknown-linux-gnu".
...
```
This is especially useful for embedded targets, e.g. arm-none-eabi-gdb.
The only annoying bit is that although GDB can normally target multiple platforms (`--enable-targets=,,`), this cross GDB can only debug one target... might just be my ignorance of Autoconf though.

(As a side note... any reason why this can't be generalized to `lib.platforms.unix`? It's obviously broadening the advertised support, but it seems the BSDs support GDB just fine...)


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Feyorsh committed:
gdb: support aarch64-darwin as a hostPlatform

GDB can still be used for cross debugging. This is useful for
e.g. arm-none-eabi-gdb.

SuperSandro2000 commented on pkgs/development/tools/misc/gdb/default.nix:
```diff
@@ -27,6 +27,7 @@ let
 in
 
 assert pythonSupport -> python3 != null;
+assert lib.asserts.assertMsg (stdenv.targetPlatform.system != "aarch64-darwin") "GDB does not support aarch64-darwin as a target";
```
Can't this use meta.badPlatforms?

Feyorsh replied:
I don't believe so; the code for checking against `platforms` and `badPlatforms` appears to use `hostPlatform`:
https://github.com/NixOS/nixpkgs/blob/40222c1c96c73fdb47070417a5ae34ffb314ad35/pkgs/stdenv/generic/check-meta.nix#L104-L105

Is there a better way to advertise support for a `hostPlatform`, but not a `targetPlatform`?

Feyorsh commented:
cc @nbp @globin @lsix, don't know why ofborg won't request review

emilazy reviewed (approved):
Looks good, thanks!

I think `lib.platforms.unix` sounds reasonable, so feel free to open another PR for that.

emilazy commented:
I’ll merge after you resolve the conflicts.

Feyorsh commented:
> I think `lib.platforms.unix` sounds reasonable, so feel free to open another PR for that.

```
nix-repl> :a pkgs.lib 
Added 412 variables.

nix-repl> :a pkgs.lib.platforms 
Added 40 variables.

nix-repl> subtractLists (linux ++ cygwin ++ freebsd ++ darwin) unix                    
[ "x86_64-solaris" "aarch64-netbsd" "armv6l-netbsd" "armv7a-netbsd" "armv7l-netbsd" "i686-netbsd" "m68k-netbsd" "mipsel-netbsd" "powerpc-netbsd" "riscv32-netbsd" "riscv64-netbsd" "x86_64-netbsd" "i686-openbsd" "x86_64-openbsd" "x86_64-redox" ]
```
Looking at https://github.com/NixOS/nixpkgs/pull/328782, I actually think support for the BSDs is a little more involved than I previously thought (I also don't have access to any of these platforms). If someone wants to extend support in the future it probably won't be too difficult.

emilazy commented:
Makes sense. I don’t think Nix on non‐FreeBSD BSDs is much of a going concern right now, anyway.

oxalica commented:
This PR breaks `availableOn` check for whether gdb is supported, making it hard to "enable gdb if the platform support it" without `tryEval`.

Reproduce:
`nix-instantiate --eval --expr 'with import ./. { system="aarch64-darwin"; }; lib.meta.availableOn stdenv.hostPlatform gdb'`

It used to return `false` but now abort with an assertion failure.

Discovered on CI: https://github.com/oxalica/nil/actions/runs/10650054468/job/29520968679

emilazy commented:
Hmm, maybe we should use `broken` instead? Or even condition `platforms` based on what `stdenv.targetPlatform` is? That feel gross, but…

Feyorsh commented:
Good catch @oxalica!
I now understand what Sandro was talking about with `meta.badPlatforms` (CUDA does something similar), so I switched to that instead. I'll spin off a PR fixing this.
